velox/functions/sparksql/String.h

```diff
@@ -293,6 +292,102 @@ struct EndsWithFunction {
   }
 };
 
+/// locate function
+/// locate(substring, string, start) -> integer
+///
+/// Returns the 1-based position of the first occurrence of 'substring' in
+/// 'string' after the give 'start' position. The search is from the beginning
+/// of 'string' to the end. 'start' is the starting character position in
+/// 'string' to search for the 'substring'. 'start' is 1-based and must be at
+/// least 1 and at most the characters number of 'string'.
+///
+/// The following rules on special values are applied to follow Spark's
+/// implementation. They are listed in order of priority:
+/// Returns 0 if 'start' is NULL. Returns NULL if 'substring' or 'string' is
+/// NULL. Returns 0 if 'start' is less than 1. Returns 1 if 'substring' is
+/// empty. Returns 0 if 'start' is greater than the characters number of
+/// 'string'. Returns 0 if 'substring' is not found in 'string'.
+template <typename T>
+struct LocateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<int32_t>& result,
+      const arg_type<Varchar>& subString,
+      const arg_type<Varchar>& string,
+      const arg_type<int32_t>& start) {
+    if (start < 1) {
+      result = 0;
+    } else if (subString.empty()) {
+      result = 1;
+    } else if (start > string.size()) {
+      result = 0;
+    } else {
+      const auto position = stringImpl::stringPosition<true /*isAscii*/>(
+          std::string_view(
+              string.data() + start - 1, string.size() - start + 1),
+          std::string_view(subString.data(), subString.size()),
+          1 /*instance*/);
+      if (position) {
+        result = position + start - 1;
+      } else {
+        result = 0;
+      }
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<int32_t>& result,
+      const arg_type<Varchar>* subString,
+      const arg_type<Varchar>* string,
+      const arg_type<int32_t>* start) {
+    if (start == nullptr) {
+      result = 0;
+      return true;
+    }
+    if (subString == nullptr || string == nullptr) {
+      return false;
+    }
+    if (*start < 1) {
+      result = 0;
+      return true;
+    }
+    if (subString->empty()) {
+      result = 1;
+      return true;
+    }
+    if (*start > stringImpl::length<false /*isAscii*/>(*string)) {
+      result = 0;
+      return true;
+    }
+
+    // Find the start byte index of the start character. For example, in the
+    // Unicode string "😋😋😋", each character occupies 4 bytes. When 'start' is
+    // 2, the 'startByteIndex' is 4 which specifies the start of the second
+    // character.
+    const char* pos = string->data();
```

Comment:
I thought we have a helper function to find the location of the N-th char.

Reply:
Yes, I found the helper function 'stringCore::cappedByteLengthUnicode' and changed to use it. Thanks for the remainder.